Gym/FrozenLake.swift

```diff
@@ -0,0 +1,156 @@
+// Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Python
+import TensorFlow
+
+let np = Python.import("numpy")
+let gym = Python.import("gym")
+
+/// Solves the FrozenLake RL problem via Q-learning. This model does not use a
+/// neural net, and instead demonstrates Swift host-side numeric processing as
+/// well as Python integration.
+
+let discountRate: Float = 0.9
+let learningRate: Float = 0.2
+let testEpisodeCount = 20
+
+typealias State = Int
+typealias Action = Int
+
+// Force unwrapping with ! does not provide source location when unwrapping
+// nil, so we instead make a util function for debuggability.
+fileprivate extension Optional {
+    func unwrapped(file: StaticString = #file, line: UInt = #line) -> Wrapped {
+        guard let unwrapped = self else {
+            fatalError("Value is nil", file: file, line: line)
+        }
+        return unwrapped
+    }
+}
+
+// This struct is defined so that `StateAction` can be a dictionary key
+// type. Swift tuples cannot be dictionary key types.
+struct StateAction: Equatable, Hashable {
+    let state: State
+    let action: Action
+}
+
+// Comparison function for conforming to Equatable protocol.
+func == (lhs: StateAction, rhs: StateAction) -> Bool {
+    return lhs.state == rhs.state && lhs.action == rhs.action
+}
+
+class Agent {
+    /// The number of actions.
+    let actionCount: Int
+    /// The current training environmental state that the agent is in.
+    var state: State
+    /// The "action value" (expected future reward value) of a pair of state and action.
+    var values: [StateAction: Float] = [:]
```

Comment:
"Value" is a very abstract term in Swift and it is completely unclear at use sites. Consider renaming this to `actionValues`?

```suggestion
    var actionValues: [StateAction: Float] = [:]
```

Reply:
done